save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt

```diff
@@ -7,100 +7,156 @@
+import io.swagger.v3.oas.annotations.enums.ParameterIn
+import io.swagger.v3.oas.annotations.responses.ApiResponse
+import io.swagger.v3.oas.annotations.tags.Tag
+import io.swagger.v3.oas.annotations.tags.Tags
+import org.springframework.security.access.prepost.PreAuthorize
 import org.springframework.security.core.Authentication
 import org.springframework.web.bind.annotation.*
-import org.springframework.web.server.ResponseStatusException
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+import reactor.kotlin.core.util.function.component1
+import reactor.kotlin.core.util.function.component2
 
 /**
  * Controller for processing links between users and their roles in projects
  */
+@ApiSwaggerSupport
+@Tags(
+    Tag(name = "projects"),
+)
 @RestController
 @RequestMapping("/api/$v1/projects/")
 class LnkUserProjectController(
     private val lnkUserProjectService: LnkUserProjectService,
     private val projectService: ProjectService,
     private val projectPermissionEvaluator: ProjectPermissionEvaluator,
 ) {
-    private val logger = LoggerFactory.getLogger(LnkUserProjectController::class.java)
-
-    /**
-     * @param organizationName
-     * @param projectName
-     * @param authentication
-     * @return list of users with their roles, connected to the project
-     * @throws NoSuchElementException
-     * @throws ResponseStatusException
-     */
     @GetMapping(path = ["/{organizationName}/{projectName}/users"])
+    @RequiresAuthorizationSourceHeader
+    @PreAuthorize("permitAll()")
+    @Operation(
+        method = "GET",
+        summary = "Get users from project with their roles.",
+        description = "Get list of users that are connected with given project and their roles in it.",
+    )
+    @Parameters(
+        Parameter(name = "organizationName", `in` = ParameterIn.PATH, description = "name of an organization", required = true),
+        Parameter(name = "projectName", `in` = ParameterIn.PATH, description = "name of a project", required = true),
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully fetched users from project.")
+    @ApiResponse(responseCode = "404", description = "Project with such name was not found.")
     fun getAllUsersByProjectNameAndOrganizationName(
         @PathVariable organizationName: String,
         @PathVariable projectName: String,
         authentication: Authentication,
-    ): List<UserInfo> {
-        val project = projectService.findByNameAndOrganizationName(projectName, organizationName)
-            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
-        val usersWithRoles = if (projectPermissionEvaluator.hasPermission(authentication, project, Permission.READ)) {
+    ): Mono<List<UserInfo>> = Mono.zip(
+        Mono.just(organizationName),
+        Mono.just(projectName),
+    )
+        .flatMap {
+            projectService.findByNameAndOrganizationName(projectName, organizationName).toMono()
+        }
```

Comment:
Why not start with simple `Mono.fromCallable { projectService... }`? You are not using parameters passed down from `zip`

Reply:
As far as I remember, @nulls considered it to be more readable, but maybe I got it wrong

Reply:
Fixed it where suitable

Reply:
I suggest to use `.toMono` for nullable values